heartbeat/monitors/wrappers/monitorstate/monitorstate.go

```diff
@@ -0,0 +1,159 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package monitorstate
+
+import (
+	"fmt"
+	"time"
+)
+
+// FlappingThreshold defines how many consecutive checks with the same status
+// must occur for us to end a flapping state. FlappingThreshold-1 is the number
+// of consecutive checks that is insufficient to start a new state, but rather to
+// keep the current state and turn it into a flapping state.
+const FlappingThreshold = 3
+
+type StateStatus string
+
+const (
+	StatusUp       StateStatus = "up"
+	StatusDown     StateStatus = "down"
+	StatusFlapping StateStatus = "flap"
+)
+
+func newMonitorState(monitorId string, status StateStatus, ctr int) *State {
+	now := time.Now()
+	ms := &State{
+		// ID is unique and sortable by time for easier aggregations
+		// Note that we add an incrementing counter to help with the fact that
+		// millisecond res isn't quite enough for uniqueness (esp. in tests)
+		ID:         fmt.Sprintf("%s-%x-%x", monitorId, now.UnixMilli(), ctr),
+		StartedAt:  now,
```

Comment:
I'm not sure that "now" is the right started at in all cases. 

Consider the complexity introduced with flapping:

Let's say a monitor runs every 5 minutes, went down at 10:30 and came back up at 10:35. Because of the flap threshold, we'd wait another 2 checks, until 10:45, to consider this down state closed.

In actuality, the down event only occurred between 10:30 and 10:35. 

That means the next state actually begins at 10:35, not 10:45, when we finally determined that the monitor was indeed NOT flapping. This function is called from `transitionTo`, which is only called after the flapping threshold of checks is complete. Because of this, the new state would register as starting at 10:45, not 10:35, leaving a gap in states of 10 minutes.

Once we determine that the monitor is indeed not flapping, it seems we need to backdate the new state to start at the first stable check (ex: 10:35), and include first `FLAPPING_THRESHOLD` number of checks as part of the history of the new state.

Reply:
That's a good point, and it's a limitation of this flapping implementation. For a short incident the flapping check does the wrong thing.

It might be best to disable flapping for now (we could just set a flap threshold of 1) and do a follow-up PR to address flapping. It's a complex topic, but I don't want to hold this up for that.

There's a choice in here in terms of whether we do want to backdate / rewrite history. The current code makes the choice to never do that. That could create a lot of confusing UX. I imagine in a UI it'd be confusing to see things shift in time etc. I'd prioritize stability of events over accuracy if that makes sense. I'm open to discussion here however.

Another option would be to require at least 3 state changes within a period of time to enter flapping. A quick outage and recovery should probably never result in a flap.

Reply:
I think it's best to disable flapping for now and do a follow up PR, given it's complexity

> That could create a lot of confusing UX.

I think that there are UX challenges on both sides of this conversation. I think the complexity of flapping currently could also cause a confusing, and potentially inaccurate UX. If we can't determine if the state is "closed" until we've ruled out flapping, then the duration of the state always has the potential to be `flappingThreshold * monitorSchedule` greater than the actual duration of the error event. I think this would be even more confusing for users, to see down events that are longer than their systems were actually down.

With the flapping complexity, the `ends` optimization becomes less helpful. It would require querying against `ends`, then checking if the last `flappingThreshold` number of checks was consistent. If so, consider the event duration as `duration - (flappingThreshold * schedule)`. 

I want to avoid bloating the timings of events when displaying in the UI as much as possible, to give our users better fidelity of the actual duration and timing of error events. 

So I think it's best to table flapping for now, and continue thinking through the complexity at another time.